src/share/index_usage/ob_index_usage_info_mgr.cpp

```diff
@@ -0,0 +1,219 @@
+                                     OB_MALLOC_NORMAL_BLOCK_SIZE, attr))) {
+    LOG_WARN("init allocator failed", K(ret));
+  } else {
+    report_task_.sql_proxy_ = GCTX.sql_proxy_;
+    report_task_.mgr_ = this;
+    is_inited_ = true;
+  }
+  return ret;
+}
+
+void ObIndexUsageInfoMgr::destroy() {
+  if (is_inited_) {
+    // cancel report task
+    if (report_task_.is_inited_) {
+      bool is_exist = true;
+      if (TG_TASK_EXIST(MTL(omt::ObSharedTimer*)->get_tg_id(), report_task_, is_exist)==OB_SUCCESS && is_exist) {
+        TG_CANCEL_TASK(MTL(omt::ObSharedTimer*)->get_tg_id(), report_task_);
+        TG_WAIT_TASK(MTL(omt::ObSharedTimer*)->get_tg_id(), report_task_);
+        report_task_.is_inited_ = false;
+      }
+    }
+    // destroy hashmap
+    index_usage_map_.destroy();
+  }
+}
+
+int ObIndexUsageInfoMgr::start(){
+  int ret = OB_SUCCESS;
+  if (is_inited_) {
+    if (OB_FAIL(TG_SCHEDULE(MTL(omt::ObSharedTimer*)->get_tg_id(),
+                                 report_task_, 
+                                 INDEX_USAGE_REPORT_INTERVAL,
+                                 true))) {
+      LOG_WARN("failed to schedule index usage report task", K(ret));
+    } else {
+      report_task_.is_inited_ = true;
+    }
+  }
+  return ret;
+}
+
+void ObIndexUsageInfoMgr::stop(){
+  if (OB_LIKELY(report_task_.is_inited_)) {
+    TG_CANCEL_TASK(MTL(omt::ObSharedTimer*)->get_tg_id(), report_task_);
+  }
+}
+
+void ObIndexUsageInfoMgr::wait(){
+  if (OB_LIKELY(report_task_.is_inited_)) {
+    TG_WAIT_TASK(MTL(omt::ObSharedTimer*)->get_tg_id(), report_task_);
+  }
+}
+
+int ObIndexUsageInfoMgr::update(const uint64_t tenant_id,
+                                const uint64_t table_id,
+                                const uint64_t index_table_id) {
+  int ret = OB_SUCCESS;
+  if (!GCONF._iut_enable || !is_inited_) {
+    // do nothing
+  } else {
+    ObIndexUsageKey key(tenant_id, table_id, index_table_id);
+    ObIndexUsageOp update_op(ObIndexUsageOpMode::UPDATE);
+    if (OB_SUCC(index_usage_map_.atomic_refactored(key, update_op))) {
+      // key exists, update success
+    } else if (OB_LIKELY(ret == OB_HASH_NOT_EXIST)) {
+      // key not exist, insert new one
+      ObIndexUsageInfo new_info(index_table_id);
+      if (GCONF._iut_max_entries <= index_usage_map_.size()) {
+        ret = OB_ERROR;
+        LOG_WARN("index usage hashmap reach max entries", K(ret));
+      } else if (OB_FAIL(index_usage_map_.set_or_update(key, new_info, update_op))) {
+        LOG_WARN("failed to set or update index-usage map", K(ret));
+      }
+    } else {
+      LOG_WARN("failed to update index-usage map", K(ret));
+    }
+  }
+  return ret;
+}
+
+/*
+1. sample ObIndexUsageInfo and call @update_func to process them by batch size SAMPLE_BATCH_SIZE
+2. check deleted index and call @del_func to del them
+*/
+int ObIndexUsageInfoMgr::sample(const UpdateFunc& update_func, const DelFunc& del_func) {
```

Comment:
It is recommended to change it according to this format like: 
if ( index table not exist ) {
       // delete record
} else if ( batch_full or loop end) {
      // update
      // clear pair_list
} else {
      // push to pair_list
}

Reply:
now we use:
if ( index table not exist ) {
// delete record
} else if (not sampled) {
// skip
} else {
// push to pair_list
}

if (batch_full or loop end) {
// update
// clear pair_list
}